src/Components/TestExplorer.fs

```diff
@@ -34,411 +165,797 @@ type TestItem with
+                    if projectRunRequest.HasIncludeFilter then
+                        Some(buildFilterExpression projectRunRequest.Tests)
+                    else
+                        None
+
+
+                let! trxPath, output = DotnetTest.runTests projectPath filterExpression
+
+                TestRun.appendOutputLine testRun output
+
+                let testResults =
+                    TrxParser.extractTrxResults trxPath |> Array.map trxResultToTestResult
+
+                if Array.isEmpty testResults then
+                    let message =
+                        $"WARNING: No tests ran for project \"{projectPath}\". \r\nThe test explorer might be out of sync. Try running a higher test or refreshing the test explorer"
+
+                    window.showWarningMessage (message) |> ignore
+                    TestRun.appendOutputLine testRun message
+                else
+                    mergeResultsToExplorer testRun projectPath runnableTests testResults
+        }
+
+
+    let private filtersToProjectRunRequests (rootTestCollection: TestItemCollection) (runRequest: TestRunRequest) =
+        let testSelection =
+            runRequest.``include``
+            |> Option.map Array.ofSeq
+            |> Option.defaultValue (rootTestCollection.TestItems())
+
+        testSelection
+        |> Array.groupBy (fun t -> TestItem.getProjectPath t.id)
+        |> Array.map (fun (projPath: string, tests) ->
+            { ProjectPath = projPath
+              HasIncludeFilter = Option.isSome runRequest.``include``
+              Tests = tests })
+
+    let runHandler
+        (testController: TestController)
+        (tryGetLocation: TestId -> LocationRecord option)
+        (req: TestRunRequest)
+        (_ct: CancellationToken)
+        : U2<Thenable<unit>, unit> =
+
+        let testRun = testController.createTestRun req
+
+        logger.Debug("TestRunRequest", req)
+
+        if testController.items.size < 1. then
+            !! testRun.``end`` ()
+        else
+            let projectRunRequests = filtersToProjectRunRequests testController.items req
+
+            logger.Debug("Project run requests", projectRunRequests)
+
+            let testItemFactory = TestItem.itemFactoryForController testController
+
+            let mergeTestResultsToExplorer =
+                mergeTestResultsToExplorer testController.items testItemFactory tryGetLocation
+
+            let runTestProject = runTestProject mergeTestResultsToExplorer testRun
+
+            promise {
+                let! _ = projectRunRequests |> Array.map runTestProject |> Promise.all
+                testRun.``end`` ()
+            }
+            |> (Promise.toThenable >> (!^))
+
+
+    let refreshTestList testItemFactory (rootTestCollection: TestItemCollection) tryGetLocation =
+        promise {
+            let! _ =
+                Project.getAll ()
+                |> List.map (fun p -> DotnetTest.dotnetTest p [||])
+                |> Promise.Parallel
```

Comment:
Yeah these seems to be the culprit. However, turning this sequential seems to also be quite slow on FAKE. Maybe we need to be smarter here and somehow send from FSAC what we think the test projects are or some other way to determine it. Thoughts @baronfel?

Reply:
Yes - FSAC should be able to help here quite a bit. Since this tooling now relies on the Test SDK (to power the `dotnet test` integration, test list, etc) FSAC should learn how to tell you if a project references that package. Consider using `Project.getInWorkspace()` instead (which also makes sure that you're only loading projects in the solution that the user has selected), which will give you a `ProjectLoadingState seq`. From this you can filter to those items that are the `ProjectLoadingState.Loaded` DU case, which has a `Project` field - this `Project` has the `PackageReference` array , and you can look for a `PackageReference` to `Microsoft.NET.Test.Sdk` to get your answer.

Reply:
Good finds. I'll see what I can figure out for this and the missing FsToolkit.ErrorHandling tests.

I did consider several alternative approaches to test discovery. The two that seemed plausible
- A test project name pattern setting 
- Build in project graph order (investigated, but I wasn't able to get that order easily).

Unfortunately, selecting test projects based on the code-discovered tests would miss test projects in other languages, unknown frameworks, or that only use Expecto test methods by proxy. 

Trying `dotnet test --list-tests` doesn't work because we're not guaranteed fully qualified test names (i.e. MSTest lists method name only). It probably wouldn't solve this issue anyway, since each project still needs to build before the tests can be listed.

Reply:
> I'll see what I can figure out for this and the missing FsToolkit.ErrorHandling tests.

I did do a full rebuild and forced a refresh of the test list and it seemed to work at that point. I'm guessing it was failing on the generation of the `trx` file since it uses `--no-build`. I wonder if there's a way to report that we couldn't get a trx file out of a project so it's more clear to a user that something else needs to happen.

Reply:
I experimented and think the missing tests and failure to load any tests are both related to parallel builds causing each other to error. I ran into some issues trying to fix this and it may take a bit to figure out.

First, I can't look for `Microsoft.NET.Test.Sdk` because projects using Paket aren't listing their package dependencies, so checking for that dependency causes no tests to be found.

Second, I tried to build all projects in sequence then test in parallel, but I'm getting unclear build errors. Still trying to figure out why this is failing.

Reply:
@farlee2121 the comment about Paket _shouldn't_ be the case - after a Restore has occured Paket still generates PackageReference MSBuild items that can be inspected and returned by FSAC. This is how the 'Package Dependencies' node in the solution explorer works for Paket-using projects. You can see this in action in `obj/<projectName>.paket.props` in a given project, for example.

Reply:
Aha! Restore was what I overlooked. I was filtering projects before restore was run.

I think initial test discovery is stable now. I'm still trying to test FAKE. Even loading the workspace is hard for my computer.

Analyzing package dependencies also opens up a new possibility for speeding up discovery. We can look for test libraries we know list full names with `--lists-tests` to avoid a full test run. The issue is projects with multiple test libraries, though any quirks should resolve with the first test run.

FsToolkit.ErrorHandling revealed another overlooked issue: multiple target framework versions. This explorer is only showing test results for one version while `dotnet test` and Visual Studio show a test item for each framework version.

Reply:
Don't stress about TFM evaluation too strongly for this iteration - we have a number of issues already around the selected project context. For example, we always assume the `DEBUG` configuration right now. TFM is just another instance of this. IIRC we use the 'first' TFM in a TFM list. You should feel free to keep this convention and we should work towards a better model all-up.